question_answering_based_on_context_using_openai_and_pinecone/main.bal

```diff
@@ -0,0 +1,95 @@
+final text:Client openAIText = check new ({auth: {token: openAIToken}});
+final embeddings:Client openaiEmbeddings = check new ({auth: {token: openAIToken}});
+final pinecone:Client pineconeClient = check new ({apiKey: pineconeKey}, serviceUrl = pineconeServiceUrl);
+
+const MAXIMUM_NO_OF_DOCS = 10;
+
+service / on new http:Listener(8080) {
+    function init() returns error? {
+        sheets:Range range = check gSheets->getRange(sheetId, sheetName, "A2:B");
+        pinecone:Vector[] vectorArray = [];
+
+        foreach any[] row in range.values {
+            string title = <string>row[0];
+            string content = <string>row[1];
+            float[] vector = check getEmbedding(string `${title} ${"\n"} ${content}`);
+            vectorArray[vectorArray.length()] = {id: title, values:vector, metadata: {"content": content}};
+        }
+
+        pinecone:UpsertRequest req = {vectors: vectorArray, namespace: "ChoreoDocs"};
+        pinecone:UpsertResponse response = check pineconeClient->/vectors/upsert.post(req);
+
+        if response.upsertedCount != range.values.length() {
+            return error("Failed to insert embedding vectors to pinecone.");
+        }
+        io:println("Successfully inserted embedding vectors to pinecone.");
+    }
+
+    resource function get answer(string question) returns string?|error {
+        string prompt = check constructPrompt(question);
+        text:CreateCompletionRequest prmt = {
+            prompt: prompt,
+            model: "text-davinci-003",
+            max_tokens: 2000
+        };
+        text:CreateCompletionResponse completionRes = check openAIText->/completions.post(prmt);
+        return completionRes.choices[0].text;
+    }
+}
+
+function getEmbedding(string text) returns float[]|error {
+    embeddings:CreateEmbeddingRequest embeddingRequest = {
+        input: text,
+        model: "text-embedding-ada-002"
+    };
+    embeddings:CreateEmbeddingResponse embeddingRes = check openaiEmbeddings->/embeddings.post(embeddingRequest);
+    return embeddingRes.data[0].embedding;
+}
+
+function countWords(string text) returns int => regex:split(text, " ").length();
+
+function constructPrompt(string question) returns string|error {
+    float[] questionEmbedding = check getEmbedding(question);
+
+    pinecone:QueryRequest req = {namespace: "ChoreoDocs", topK: MAXIMUM_NO_OF_DOCS, vector: questionEmbedding, includeMetadata: true};
+    pinecone:QueryResponse res = check pineconeClient->/query.post(req);
+    pinecone:QueryMatch[]? rows = res.matches;
+
+    string context = "";
+    int contextLen = 0;
+    int maxLen = 1125; // approx equivalence between word and token count
+
+    if rows is pinecone:QueryMatch[] {
+        foreach pinecone:QueryMatch row in rows {
+            pinecone:VectorMetadata? rowMetadata = row.metadata;
+            if rowMetadata is pinecone:VectorMetadata {
+                string content =  check rowMetadata["content"].ensureType();
```

Comment:
There was a request to not use `ensureType`, right?

Reply:
I think we have to do a type casting here since `rowMetadata["content"]` has the type `anydata`. Let me know if there is a way to avoid using `ensureType`.